### ISSUE TITLE
protobuf-net.Grpc  Emit  support xamarin error

### ISSUE THREAD

**Issue by neshcodes** [protobuf-net/protobuf-net.Grpc#65]
{System.TypeInitializationException: The type initializer for 'DefaultProxyCache`3' threw an exception. ---> System.TypeInitializationException: The type initializer for 'ProtoBuf.Grpc.Internal.ProxyEmitter' threw an exception. ---> System.PlatformNotSupportedException: Operation is not supported on this platform.
  at System.Reflection.Emit.AssemblyBuilder.DefineDynamicAssembly (System.Reflection.AssemblyName name, System.Reflection.Emit.AssemblyBuilderAccess access) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/src/Xamarin.iOS/mcs/class/corlib/System.Reflection.Emit/AssemblyBuilder.pns.cs:129

**Comment by mgravell** [comment]
And what operation are you performing here? Is this creating a client
proxy, to be clear?

**Comment by neshcodes** [comment]
just test the code, test example on the xamarin error, but on the netcore console client is ok

**Comment by mgravell** [comment]
The client part of the tooling here isn't going to work on AOT in the
current form. That is something we could revisit when "generators" lands
(Roslyn), or we could look at other options for emitting a client during
build.

Are you purely code-first? Or are you using protogen to emit the interface?
We could emit a client proxy class from .proto, for example.

**Comment by neshcodes** [comment]
i use code-first, https://github.com/neshcodes/grpc_xam_test/blob/master/GrpcApp.Server/Startup.cs an in this demo, i cant found grpc func in this new-demo:
warn: Grpc.AspNetCore.Server.Model.Internal.ServiceRouteBuilder[3]
      No gRPC methods discovered for GrpcApp.Server.Services.GreeterService.
warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Overriding address(es) 'https://localhost:5001'. Binding to endpoints defined in UseKestrel() instead.

**Comment by mgravell** [comment]
I'll look into the Kestrel bit. Having a GitHub to pull down helps a lot,
ta.

On the client: crazy idea, but: if AOT won't work today, would "dynamic"
suffice? Meaning: you wouldn't have intellisence, but if method worked...?

**Comment by neshcodes** [comment]
i`m testing in android with .proto

**Comment by mgravell** [comment]
Re AOT: the other thing we could do, if `dynamic` isn't a great fit, is to create a universal proxy, something like [here](https://github.com/protobuf-net/protobuf-net.Grpc/blob/7220cc879363fb234874e674dd24c9832c244844/tests/protobuf-net.Grpc.Test.Integration/GrpcServiceTests.cs#L93-L110), but in the main library and with support for all the calling conventions, not just "unary"; what we'd need to figure out is the signature/usage; getting the *client* is easy enough - we could do something like:

``` c#
// note NONE OF THIS EXISTS TODAY

// use a new non-generic API to get a universal proxy; this resolves
// the service name and prepares some internal state
var client = channel.CreateGrpcService(typeof(IGreeterService));
```

but the question is the usage; i.e. is it:

``` c#
// simple to implement
var response = await client.AsyncUnaryCall<HelloRequest, HelloReply>(
    request, nameof(IGreeterService.SayHello));
```

or something more esoteric, perhaps using expression trees, i.e. something like

``` c#
var response = await client.AsyncUnaryCall(svc => svc.SayHello(request));
```

**Comment by mgravell** [comment]
FWIW: I suspect either of the approaches directly above would be better than `dynamic`, so: I'm tempted to say:

1. ditch the `dynamic` idea
2. implement the simple version ASAP
3. leave a place-holder to investigate the expression-tree version

(note: the code for the expression-tree version isn't particularly gnarly - I've done exactly that before  for protobuf-net way way back; it is just more nuanced re testing etc)

**Comment by mgravell** [comment]
---

Re the Kestrel bit; the problem is that `HelloRequest` isn't marked as a data-contract (`[DataContract]`); this meant that it couldn't resolve the signature correctly; I will see if I can improve the logging of this in the Kestrel output.

**Comment by mgravell** [comment]
Applying a logging fix for this; new output:

``` txt
warn: ProtoBuf.Grpc.Server.ServicesExtensions.CodeFirstServiceMethodProvider[0]
      Type cannot be serialized; ignoring: GrpcApp.Common.HelloRequest
warn: ProtoBuf.Grpc.Server.ServicesExtensions.CodeFirstServiceMethodProvider[0]
      Signature not recognized for 'SayHello'; method will not be bound
warn: Grpc.AspNetCore.Server.Model.Internal.ServiceRouteBuilder[3]
      No gRPC methods discovered for GrpcApp.Server.Services.GreeterService.
```

**Comment by mgravell** [comment]
The improved logging is now available in 1.0.26

**Comment by neshcodes** [comment]
cause i made a stupid mistake

**Comment by neshcodes** [comment]
the console client run success, the xamarin android client log : Grpc.Core.RpcException: Status(StatusCode=Internal, Detail="Error starting gRPC call: Failed to connect to localhost/127.0.0.1:50001",

**Comment by neshcodes** [comment]
so i need add a proxy?

**Comment by mgravell** [comment]
by "proxy" here, I'm talking about the client proxy; *currently*, when you call:

``` c#
channel.CreateGrpcService<IGreeterService>();
```

behind the scenes, it basically does:

1. have I seen `IGreeterService` before? if so: jump to step 4
2. what does `IGreeterService` look like?
3. generate a new type on the fly that implements `IGreeterService` but knows how to connect the API through to the gRPC mechanisms
4. create an instance of that type against the specified channel, and hand it back to the client

Now; the problem is: step 3 **cannot work** in a pure AOT scenario; so what I'm saying is: if you need this in AOT, we need to think of different options; a few leap to mind:

1. use "generators" to perform steps 2 & 3 *during build*; this is the ideal solution *for everyone*, but the slight problem is that "generators" hasn't shipped yet
2. use some other build step to perform steps 2 & 3; but the problem is where that info comes from; if we were using "protogen", there might be options, but for code-first; not so much
3. use some kind of universal proxy that doesn't actually implement `IGreeterService` itself, but which makes it possible to access the *methods* of `IGreeterService`

What I was discussing [in this comment](https://github.com/protobuf-net/protobuf-net.Grpc/issues/65#issuecomment-592390919) was a couple of different options for implementing option 3. Happy to do the work there (it isn't actually a lot to do) - we just need to understand what will be useful.

**Comment by mgravell** [comment]
Ultimately, what I'm saying is: until we get "generators", `channel.CreateGrpcService<IGreeterService>` cannot work in a pure AOT scenario. Happy to revisit that when "generators" lands, but there is no ETA on that. So in the interim; can we discuss an alternative API that will allow you to do what you need; for example:

``` c#
var client = channel.CreateUniversalGrpcService<IGreeterService>();
// option A
var response = await client.AsyncUnaryCall<HelloRequest, HelloReply>(
    request, nameof(IGreeterService.SayHello));
// option B
var response = await client.AsyncUnaryCall(svc => svc.SayHello(request));
```

**Comment by neshcodes** [comment]
got it ,  thanks for your answer

**Comment by neshcodes** [comment]
i add ext with public static class GrpcExtensions
    {
        public static Task<TResponse> Execute<TRequest, TResponse>(this GrpcChannel channel, TRequest request, string serviceName, string methodName,
            CallOptions options = default, string? host = null)
            where TRequest : class
            where TResponse : class
            => Execute<TRequest, TResponse>(channel.CreateCallInvoker(), request, serviceName, methodName, options, host);

        public static async Task<TResponse> Execute<TRequest, TResponse>(this CallInvoker invoker, TRequest request, string serviceName, string methodName,
            CallOptions options = default, string? host = null)
            where TRequest : class
            where TResponse : class
        {
            var method = new Method<TRequest, TResponse>(MethodType.Unary, serviceName, methodName,
                CustomMarshaller<TRequest>.Instance, CustomMarshaller<TResponse>.Instance);
            using (var auc = invoker.AsyncUnaryCall(method, host, options, request))
            {
                return await auc.ResponseAsync;
            }
        }

        class CustomMarshaller<T> : Marshaller<T>
        {
            public static readonly CustomMarshaller<T> Instance = new CustomMarshaller<T>();
            private CustomMarshaller() : base(Serialize, Deserialize) { }

            private static T Deserialize(byte[] payload)
            {
                using (var ms = new MemoryStream(payload))
                {
                    return ProtoBuf.Serializer.Deserialize<T>(ms);
                }
            }
            private static byte[] Serialize(T payload)
            {
                using (var ms = new MemoryStream())
                {
                    ProtoBuf.Serializer.Serialize<T>(ms, payload);
                    return ms.ToArray();
                }
            }
        }
    }

but the throw Grpc.Core.RpcException: Status(StatusCode=Unimplemented, Detail="Service is unimplemented.")

**Comment by neshcodes** [comment]
how i tag the service ,https://github.com/neshcodes/grpc_xam_test/blob/master/GrpcApp.Client/Program.cs

**Comment by mgravell** [comment]
and what parameters did you use? i.e. what `serviceName` and `methodName` did you pass? They aren't necessarily what you think they are, hence why wrapping this into the library is a good idea, so the library can deal with all that noise.

**Comment by mgravell** [comment]
ah, we passed each-other there; yeah, the name isn't `IGreeterService`; one sec, I'll find out what it is, but my guess would be `GrpcApp.Common.GreeterService`

**Comment by mgravell** [comment]
yes, it is `GrpcApp.Common.GreeterService`:

``` c#
var serviceType = typeof(IGreeterService);
Assert.True(ServiceBinder.Default.IsServiceContract(serviceType, out var serviceName));
Assert.Equal("GrpcApp.Common.GreeterService", serviceName);

var method = serviceType.GetMethod(nameof(IGreeterService.SayHello))!;
Assert.True(ServiceBinder.Default.IsOperationContract(method, out var operationName));
Assert.Equal("SayHello", operationName);
```

**Comment by mgravell** [comment]
again, what I'm trying to emphasize here is that in a "real" solution, you wouldn't need to worry about any of that - the library would just make it work

**Comment by neshcodes** [comment]
its work on the grpcapp.client, but the android throw ex : Grpc.Core.RpcException: Status(StatusCode=Internal, Detail="Error starting gRPC call: Failed to connect to localhost/127.0.0.1:50001"),i also pull the code

**Comment by neshcodes** [comment]
the ios throw : Grpc.Core.RpcException: Status(StatusCode=Internal, Detail="Error starting gRPC call: The resource could not be loaded because the App Transport Security policy requires the use of a secure connection."

**Comment by neshcodes** [comment]
so the proxy erro is fix?

**Comment by mgravell** [comment]
> so the proxy erro is fix?

I'm sorry, I don't understand the question; can you rephrase?

**Comment by neshcodes** [comment]
i mean the ex "{System.TypeInitializationException: The type initializer for 'DefaultProxyCache`3' threw an exception." not throw

**Comment by mgravell** [comment]
you shouldn't be seeing anything involving `DefaultProxyCache` if you're not using `CreateGrpcService`; again - can you clarify exactly what your situation is *now*, and what you're seeing, including the exception message and stack trace if possible? I don't want to make assumptions, because it seems a lot has changed over the last while

**Comment by neshcodes** [comment]
now the client code is :
try
                {
                    GrpcClientFactory.AllowUnencryptedHttp2 = true;
                    using (var channel = GrpcChannel.ForAddress("http://10.0.2.2:50001"))
                    {
                        var login_rpc = channel.CreateGrpcService<IGreeterService>();
                        var result = await login_rpc.SayHello(new HelloRequest() { Name = Guid.NewGuid().ToString() });
                        Console.WriteLine(result.Text);
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex);
                }


{Grpc.Core.RpcException: Status(StatusCode=Internal, Detail="Error starting gRPC call: unexpected end of stream on com.android.okhttp.Address@e08769c5")
  at ProtoBuf.Grpc.Internal.Reshape.UnaryTaskAsyncImpl[TRequest,TResponse] (Grpc.Core.AsyncUnaryCall`1[TResponse] call, ProtoBuf.Grpc.Internal.MetadataContext metadata) [0x00107] in <397bf8fc0548449f92e6bb43ec90a591>:0 
  at System.Threading.Tasks.ValueTask`1[TResult].get_Result () [0x0001b] in /Users/builder/jenkins/workspace/archive-mono/2019-08/android/release/external/corefx/src/Common/src/CoreLib/System/Threading/Tasks/ValueTask.cs:813 
  at GrpcApp.MainPage+<>c.<.ctor>b__0_0 () [0x0007d] in /Users/coder/Projects/grpc_xam_test/GrpcApp/MainPage.xaml.cs:75 }

**Comment by neshcodes** [comment]
https://github.com/neshcodes/grpc_xam_test/blob/master/GrpcApp/MainPage.xaml.cs

**Comment by mgravell** [comment]
K, I'll try and look over the weekend to see what is up.

**Comment by neshcodes** [comment]
so i need test use the master code, right? i see the version not push the nuget~

**Comment by mgravell** [comment]
I haven't looked at the most recent part of this yet (the end of stream pieces), so: probably not worth you trying yet. But: the new `GrpcClient` piece is on the myget feed: https://www.myget.org/F/protobuf-net/api/v3/index.json

**Comment by neshcodes** [comment]
error Service 'Greeter' is unimplemented , what i lost?

public class GreeterService : IGreeterService
    {
        /*public ValueTask<HelloReply> SayHello(HelloRequest request)
        {
            return new ValueTask<HelloReply>(new HelloReply
            {
                Text = "Hello " + request.Name
            });
        }*/

        [Operation]
        public async ValueTask<HelloReply> SayHello(HelloRequest request)
        {
            return new HelloReply
            {
                Text = "Hello " + request.Name
            };
        }
    }

[Service("Greeter")]
    public interface IGreeterService : IGrpcService
    {
        [Operation]
        ValueTask<HelloReply> SayHello(HelloRequest request);
    }

    [DataContract, ProtoContract]
    public class HelloReply
    {
        [DataMember(Order = 1), ProtoMember(1)]
        public string Text { get; set; }
    }

    [DataContract]
    public class HelloRequest
    {
        [DataMember(Order = 1), ProtoMember(1)]
        public string Name { get; set; }
    }

**Comment by neshcodes** [comment]
or apsnet core startup error?

public class Startup
    {
        // This method gets called by the runtime. Use this method to add services to the container.
        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddCodeFirstGrpc(config =>
            {
                //config.ResponseCompressionLevel = System.IO.Compression.CompressionLevel.Optimal;
            });
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment _)
        {
            app.UseRouting();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapGrpcService<GreeterService>();
            });
        }
    }

**Comment by mgravell** [comment]
I will look as soon as I can. Is that updated on your GitHub?

**Comment by neshcodes** [comment]
i push it

**Comment by neshcodes** [comment]
u can see server and client

**Comment by mgravell** [comment]
Right; including `: IGrpcService` changes a lot of binding rules; three options:

1: when creating the `GrpcClient`, tell it `typeof(IGreeterService)` and let *it* worry about naming, or:
2: remove `: IGrpcService`, or:
3: both!

I went with option 3

Also, locally hen I use `dotnet run` on the server, it is listening on port 50001, so I changed the port binding in the client

Here's my deltas:

``` txt
     [Service("Greeter")]
-    public interface IGreeterService : IGrpcService
+    public interface IGreeterService
```

and

``` txt
                     GrpcClientFactory.AllowUnencryptedHttp2 = true;
-                    var channel = new Channel("localhost", 12345, ChannelCredentials.Insecure);
+                    var channel = new Channel("localhost", 50001, ChannelCredentials.Insecure);
                     //using (var channel = GrpcChannel.ForAddress("https://localhost:5001"))
                     //using (var channel = GrpcChannel.ForAddress("https://localhost:5001"))
                     {
                         HelloRequest request = new HelloRequest();
                         request.Name = "11111";
-                        var client = new GrpcClient(channel, "Greeter");
+                        var client = new GrpcClient(channel, typeof(IGreeterService));
```

With that, it works fine here

**Comment by neshcodes** [comment]
the xamarin.android work success~~

**Comment by neshcodes** [comment]
and i need tag the "ProtoContract" "ProtoMember"?

**Comment by mgravell** [comment]
`[ProtoContract]` and `[ProtoMember(n)]` are used by protobuf-net for serialization purposes. If you prefer, there are other ways of annotating what it needs, but the protobuf-net attributes are usually the most flexible.

If you prefer to use a different serializer: that is also possible but requires additional work to configure it.

**Comment by mgravell** [comment]
So: have we covered everything? Is there anything outstanding?

**Comment by neshcodes** [comment]
yeh, i can use it, its work good,

**Comment by neshcodes** [comment]
but the xamarin doesn't support grpc very well，the roadmap say support xamarin at the .net5...
